### PR TITLE
WIP: AST Traverse

### PR DESCRIPTION
**Pull request by rzvxa**

### CODE REVIEW

**Comment by rzvxa** [comment]
* **#2987** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/2987?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rzvxa and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by milesj** [comment]
@rzvxa So how is this working exactly? Does `#[ast_node]` mutate the current AST node? Assumed we would create a duplicate node, like `TraversableProgram`.

**Comment by rzvxa** [comment]
> @rzvxa So how is this working exactly? Does `#[ast_node]` mutate the current AST node? Assumed we would create a duplicate node, like `TraversableProgram`.

It would modify and validate our ast_node and generate their appropriate traversable versions, mutation wrappers, etc.
It will also ensure that the 2 types have the same representation and size; It adds some constant assertions to test it out.

File: crates/oxc_ast/src/lib.rs [hunk 1]
```diff
@@ -58,15 +60,15 @@ fn size_asserts() {
     use crate::ast;
     use oxc_index::assert_eq_size;
 
-    assert_eq_size!(ast::Statement, [u8; 16]);
+    assert_eq_size!(ast::Statement, [u8; 24]);
```
[the remark this answers is not in the document]
Comment (rzvxa) [reply crates/oxc_ast/src/lib.rs:63]: changing the representation from rust to C

**Comment by overlookmotel** [comment]
I've started going through this.

So far just been tidying up a little. Have pushed a few commits to remove changes which aren't relevant to the thrust of this PR, and submitted a couple as separate PRs (#3026, #3027). Aim is to reduce the diff a little, and make it a bit easier to follow the substantive changes.

Will start going through the meat of it soon as I can.

File: crates/oxc_ast/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (overlookmotel) [reply crates/oxc_ast/src/lib.rs:63]: There are possible workarounds to reverse the type size inflation, if we want to prioritize it. It does affect the benchmarks a little (-1% on parser benchmarks), but to be honest I expected it to be much worse.

[the remark this answers is not in the document]
Comment (rzvxa) [reply crates/oxc_ast/src/lib.rs:63]: Yes, I was also expecting a hit in the performance; Not having any tangible performance drop is a sentiment to how well modern CPUs are in pre-fetching and pipelining the data.

File: crates/oxc_macros/src/ast_node.rs [hunk 2]
```diff
@@ -56,10 +56,11 @@ fn modify_enum(item: &mut ItemEnum) -> NodeData {
...
     item.variants.insert(0, parse_quote!(Dummy));
     // add explicit discriminants to all variants
-    item.variants
-        .iter_mut()
-        .enumerate()
-        .for_each(|(i, var)| var.discriminant = Some((parse_quote!(=), parse_quote!(#i as u8))));
+    item.variants.iter_mut().enumerate().for_each(|(i, var)| {
+        #[allow(clippy::cast_possible_truncation)]
+        let i = i as u8;
+        var.discriminant = Some((parse_quote!(=), parse_quote!(#i)));
```
Comment (rzvxa) [inline crates/oxc_macros/src/ast_node.rs:62]: When writing this, I checked this line, and the compiler optimizes the cast away after macro expansion; It expands to `Variant = 1 as u8` and the compiler does the cast in comptime. I've kept it that way for better readability.

Was there any issues with it? I'm just curious.

**Comment by overlookmotel** [comment]
@rzvxa It's starting to come into focus for me now.

I attempted to make it work without changing the AST further, but unfortunately can't find a way to avoid soundness issues without all enum variants being boxed, and other changes (my failed attempt to get `MemberExpression` working [here](https://github.com/overlookmotel/ast-cell-test/commit/c730ea31cd3e071abc18ec01e6a1654791ffc24a)).

My conclusion is: If there's a `visit_*` function for a type, then that type needs either:

1. to be an `enum` (which is `Copy`) or
2. to be enclosed in `Box<_>` or `Vec<_>` everywhere it's used (so that `Box<T>`/`Vec<T>` can be safely transmuted to `&GCell<T>`, which is what's passed to the `visit_*` function).

Unfortunately, that will involve quite a lot of changes to the AST.

NB: A variant on my failed attempt above should work in some cases, as long as the type is always in a fixed position in all other types (i.e. not within `Option<_>`), so a `&GCell<T>` will always point to a valid `T` forever, no matter how the AST is altered. But it's tricky, so I think better just box everything for now to get it working, and then try to scale the boxing back again later if it hurts performance.

I've added a commit to use [layout_inspect](https://github.com/overlookmotel/layout_inspect) on the AST, which allows running queries on the AST types:

```sh
node tasks/inspect_ast
```

Do have a look if you're interested.

Clocking off for tonight now, but will continue tomorrow.

**Comment by overlookmotel** [comment]
PS: I also made some bug fixes and improvements to demo impl (the last 6 commits here: https://github.com/overlookmotel/ast-cell-test/commits/main/), which we should apply to this full implementation at some point.

**Comment by overlookmotel** [comment]
All nested enums in the AST have now been squashed. So e.g. `ArrayExpressionElement` contains all the variants of `Expression`, rather than having its own variant `ArrayExpressionElement::Expression(Expression<'a>)`. This was a hard slog (4 x 14-hour days) but, aside from cleaning up the macro madness, it's done.

As a side-effect, all the type sizes are back to what they were originally, and in fact a few got smaller.

**Comment by overlookmotel** [comment]
I am pretty rinsed-out after going at it on #3115 all week. Can I ask for some help?

The last step of getting the AST in shape to be able to traverse over it is that a bunch of types need to be enclosed in `Box<_>` wherever they're used.

The types in question are:

* AssignmentTargetRest
* BindingIdentifier
* BindingPattern
* CatchParameter
* Elision
* Hashbang
* IdentifierReference
* JSXEmptyExpression
* JSXIdentifier
* LabelIdentifier
* PrivateIdentifier
* StringLiteral
* TSImportAttributes
* TSThisParameter
* TSThisType
* TemplateLiteral
* WithClause

Anywhere they appear in the AST and they're not already wrapped in `Box<_>` or `Vec<_>`, they need to be. e.g.:

```diff
pub struct ObjectAssignmentTarget<'a> {
    pub span: Span,
    pub properties: Vec<'a, AssignmentTargetProperty<'a>>,
-   pub rest: Option<AssignmentTargetRest<'a>>,
+   pub rest: Option<Box<'a, AssignmentTargetRest<'a>>>,
}
```

If anyone is willing to lend a hand, please add commits to this PR directly (if you have write access) or submit as PRs onto this branch (not main).

To get an up-to-date list of types which are not yet sorted, run on this branch:

```sh
node tasks/inspect_ast
```

**Comment by overlookmotel** [comment]
Rebased on main after #3115 merged.

**Comment by overlookmotel** [comment]
Rebased on latest main.

**Comment by overlookmotel** [comment]
A lot of work went into this, and I would like to save it somewhere. It's possible we want to come back to this later.

I'll save it in a branch and then will close again.

**Comment by rzvxa** [comment]
> A lot of work went into this, and I would like to save it somewhere. It's possible we want to come back to this later.
> 
> I'll save it in a branch and then will close again.

Do you think we could close this for housekeeping reasons? It seems like this PR has drifted too far from our current trunk, So I guess what I'm trying to ask here is do you still believe that the code here can be useful at some point?

**Comment by overlookmotel** [comment]
Yes sorry, I requested it be kept, and then forgot about it.

I've merged #3149 into this, and saved it on [traversable-ast branch](https://github.com/oxc-project/oxc/tree/traversable-ast).

I don't know for sure if we'll need it, but there was a lot of interesting stuff in this PR, so might want to go back to it/pull some stuff out if it later on.

Closing this PR.

**Pull request closed without merging**